google/cloud/datastore/helpers.py

```diff
@@ -43,36 +42,29 @@ def _get_meaning(value_pb, is_list=False):
     :param is_list: Boolean indicating if the ``value_pb`` contains
                     a list value.
 
-    :rtype: int
+    :rtype: int | Tuple[Optional[int], Optional[list[int | None]]] | None
     :returns: The meaning for the ``value_pb`` if one is set, else
-              :data:`None`. For a list value, if there are disagreeing
-              means it just returns a list of meanings. If all the
-              list meanings agree, it just condenses them.
+              :data:`None`. For a list value, returns a tuple of
+              the root meaning of the list, and a list of meanings
+              of each sub-value. If subvalues are all empty, returns
+              :data:`None` instead of a list.
     """
     if is_list:
+        root_meaning = value_pb.meaning or None
         values = value_pb.array_value.values
 
-        # An empty list will have no values, hence no shared meaning
-        # set among them.
-        if len(values) == 0:
-            return None
-
         # We check among all the meanings, some of which may be None,
         # the rest which may be enum/int values.
-        all_meanings = [_get_meaning(sub_value_pb) for sub_value_pb in values]
-        unique_meanings = set(all_meanings)
-
-        if len(unique_meanings) == 1:
-            # If there is a unique meaning, we preserve it.
-            return unique_meanings.pop()
-        else:  # We know len(value_pb.array_value.values) > 0.
-            # If the meaning is not unique, just return all of them.
-            return all_meanings
-
-    elif value_pb.meaning:  # Simple field (int32).
-        return value_pb.meaning
-
-    return None
+        sub_meanings = [sub_value_pb.meaning or None for sub_value_pb in values]
+        if not any(meaning is not None for meaning in sub_meanings):
```

Comment:
I'm curious, why wouldn't we just return a tuple of (None, []), instead of transforming the return type here to just a single None?

Reply:
In case it's not clear: the reason for this transformation is to flatten giant lists of `[None, None, None, ...]`, so we don't have to iterate over the whole thing again later

You're right that we could return an empty list instead of None here to signal that there are no meanings present. But it's a trade-off. That would buy us type consistency, but then we'd lose the property that the sub-meaning list is always be the same size as the original array. We'd need to remember to do size comparisons before iterating, so we don't really win anything from that change. And size mismatches wouldn't be flagged by type checkers like None checks, so those bugs would be harder to notice.

TL;DR: This is a special case in our processing logic, and it's usually better to use None to signal special cases, instead of returning lists with unpredictable sizes